examples/src/main/java/com/squareup/moshi/recipes/RuntimeJsonAdapterFactory.java

```diff
@@ -0,0 +1,129 @@
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonDataException;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+public final class RuntimeJsonAdapterFactory implements JsonAdapter.Factory {
+  private final Map<Class<?>, String> subtypeToLabel = new LinkedHashMap<>();
+  private final Class<?> baseType;
+  private final String labelKey;
+
+  public RuntimeJsonAdapterFactory(Class<?> baseType, String labelKey) {
+    if (baseType == null) {
+      throw new NullPointerException("baseType == null");
+    }
+    if (labelKey == null) {
+      throw new NullPointerException("labelKey == null");
+    }
+    this.baseType = baseType;
+    this.labelKey = labelKey;
+  }
+
+  public RuntimeJsonAdapterFactory registerSubtype(Class<?> subtype, String label) {
+    if (subtype == null) {
+      throw new NullPointerException("subtype == null");
+    }
+    if (label == null) {
+      throw new NullPointerException("label == null");
+    }
+    if (!baseType.isAssignableFrom(subtype)) {
+      throw new IllegalArgumentException(subtype + " must be a " + baseType);
```

Comment:
Or use generics to make this a compile-time failure. (`Class<? extends T> subtype`). Not sure.